### PR TITLE
Liked pets feature

### PR DESCRIPTION
**Pull request by kasugaijin**
# 🔗 Issue
NA

# ✍️ Description
Adds a liked pets feature so adopter users can like a pet and see them from their dashboard later. See comments below for more discussion.

https://www.loom.com/share/1d6210a54ebd40bd83c55cac955fb42a?sid=6e5bb0c8-f7e9-447c-8f47-e7bb15796a76

# 📷 Screenshots/Demos

### CODE REVIEW

File: app/controllers/organizations/adoptable_pets_controller.rb
Comment (kasugaijin) [inline app/controllers/organizations/adoptable_pets_controller.rb:6]: eager load a pet's liked pets for the index view. We need to tell a user if they have liked a pet or not, so for each pet card we will query this data to find out if their adopter_foster_account id is in the collection of the pet's liked_pets

File: app/models/concerns/authorizable.rb
Comment (kasugaijin) [inline app/models/concerns/authorizable.rb:26]: assuming there is no need for staff to like pets.

File: app/policies/organizations/liked_pet_policy.rb
Comment (kasugaijin) [inline app/policies/organizations/liked_pet_policy.rb:27]: I have left this pretty open ended such that an adopter or fosterer, even without a profile, can like and unlike any pet. I say we take this approach to start and let feedback inform us if we need to make changes. My assumption here is that people would like to be able to track liked pets before investing the time to fill out a profile. I see it as a way for users to shortlist pets before making the investment.

File: app/views/partials/_like_button.html.erb
Comment (kasugaijin) [inline app/views/partials/_like_button.html.erb:2]: Not usually a fan of doing queries in views. Not sure where this would best live, otherwise, either. If we move this method somewhere it's just going to be more code to achieve the same thing as we'd still need to pass the args to the method that then does the query...

File: app/views/partials/_pet_card.html.erb
Comment (kasugaijin) [inline app/views/partials/_pet_card.html.erb:3]: I moved the 'Adoption Pending' badge to the top right of the image so there was more room for the like functionality. I also prefer this placement of this badge to before. Thoughts?

I also make the image and the pet name clickable to go to the pet show page, instead of the whole card. This allows the user to click the like/unlike independently in the pets index view, and is consistent with how other websites link their cards. Usually details are not clickable, only names and images.

File: app/views/partials/_pet_card.html.erb
Comment (kasugaijin) [inline app/views/partials/_pet_card.html.erb:21]: I am still using this because I find it easier to read...thoughts?

File: test/policies/adopter_application_policy_test.rb
Comment (kasugaijin) [inline test/policies/adopter_application_policy_test.rb:103]: I added the adopter role to the fosterer factory seeing as we do this now.

File: test/integration/adoptable_pets_index_test.rb
Comment (kasugaijin) [inline test/integration/adoptable_pets_index_test.rb:20]: There are now two a tags because I have made the image and pet name clickable.

File: app/controllers/organizations/adoptable_pets_controller.rb
Comment (jmilljr24) [reply app/controllers/organizations/adoptable_pets_controller.rb:6]: `    @likes = LikedPet.where(adopter_foster_account: current_user&.adopter_foster_account)
`

File: app/views/partials/_pet_card.html.erb
Comment (jmilljr24) [inline app/views/partials/_pet_card.html.erb:23]:
```suggestion
              <%= render partial: 'partials/like_button', locals: { pet: pet, liked: @likes&.find { |like| like.pet_id == pet.id}  } %>
```

File: app/views/partials/_like_button.html.erb
Comment (jmilljr24) [reply app/views/partials/_like_button.html.erb:2]:
```suggestion
  <%= button_to adopter_fosterer_liked_pet_path(id: liked&.id),
```

File: app/views/organizations/adopter_fosterer/liked_pets/create.turbo_stream.erb
Comment (jmilljr24) [inline app/views/organizations/adopter_fosterer/liked_pets/create.turbo_stream.erb:3]:
```suggestion
    <%= render partial: 'partials/like_button', locals: { pet: @pet, liked:  @liked_pet } %>
```

**Review by jmilljr24 — COMMENTED** [review]
With regards to the like_button query, I think I have an alternate solution but if I'm missing the mark feel free to disregard. For clarity I'll be refer to the liked pet object from the LikedPet model not a Pet that has been 'liked' 

Starting from the adoptable_pets_controller, we could create an instance variable of all the LikedPet objects associated with the adopter.

Instead of true/false being passed to the like button, the pet_card would pass in the LikedPet object if the instances variable contains it. 

The like button would simply use the LikedPet object id for the button path instead of a query.

File: app/views/organizations/adopter_fosterer/liked_pets/create.turbo_stream.erb
Comment (kasugaijin) [reply app/views/organizations/adopter_fosterer/liked_pets/create.turbo_stream.erb:3]: Yes this will be better @jmilljr24, I will make changes.

**Comment by kasugaijin** [comment]
One thing I will also do (which I don't think is a blocker for review and approval?) is move text from views and controllers to the translations file before merging.

File: app/controllers/organizations/adopter_fosterer/liked_pets_controller.rb
Comment (kasugaijin) [inline app/controllers/organizations/adopter_fosterer/liked_pets_controller.rb:49]: I note that if I do not have these in the permitted params I get this error
```
Unpermitted parameters: :_method, :authenticity_token. Context: { controller: Organizations::AdopterFosterer::LikedPetsController, action: destroy, request: #<ActionDispatch::Request:0x000000011a825990>, params: {"_method"=>"delete", "authenticity_token"=>"[FILTERED]", "controller"=>"organizations/adopter_fosterer/liked_pets", "action"=>"destroy", "id"=>"32"} }
```
The destroy or create still works, but I added them to the permitted params list to remove the erorr. Any thoughts?

Comment (kasugaijin) [reply app/controllers/organizations/adopter_fosterer/liked_pets_controller.rb]: Yes this is the more rails way for sure.

File: package.json
Comment (mononoken) [inline package.json:1]: I think this was unintentionally generated.

File: test/factories/users.rb
Comment (mononoken) [inline test/factories/users.rb:49]: Good catch 👍

File: app/views/partials/_pet_card.html.erb
Comment (mononoken) [reply app/views/partials/_pet_card.html.erb:3]: Looks and sounds good to me! I also made the fosters cards with only the image clickable so this is consistent there.

**Review by mononoken — COMMENTED** [review]
I have not looked at the tests yet, but here is my initial review. My main concern is in the comments about the naming of `LikedPet`. Would be curious if others agree.

File: app/controllers/organizations/adopter_fosterer/liked_pets_controller.rb
Comment (kasugaijin) [reply app/controllers/organizations/adopter_fosterer/liked_pets_controller.rb:2]: Yes I agree the skip makes more sense as it occurs less often.

File: app/models/liked_pet.rb
Comment (kasugaijin) [reply app/models/liked_pet.rb:25]: 100% I agree. LikedPets is a bad name for it. Likes makes more sense.

File: app/models/adopter_foster_account.rb
Comment (kasugaijin) [reply app/models/adopter_foster_account.rb:27]: Exactly! Likes makes more sense. It even tripped me up once returning pets when I expected likes.

File: package.json
Comment (kasugaijin) [reply package.json:1]: Yes I tested out using awesome fonts. This is a remnant I can delete.

File: app/controllers/organizations/adopter_fosterer/liked_pets_controller.rb
Comment (kasugaijin) [reply app/controllers/organizations/adopter_fosterer/liked_pets_controller.rb:7]: My intention was to get this up and running then we can decide what to do here and make a future issue for it.

File: app/views/partials/_pet_card.html.erb
Comment (kasugaijin) [reply app/views/partials/_pet_card.html.erb:21]: Yes, I agree there is pros and cons to both approaches. Perhaps something to discuss broader so we can set a consistent approach through the app.

File: app/controllers/organizations/adopter_fosterer/liked_pets_controller.rb
Comment (kasugaijin) [reply app/controllers/organizations/adopter_fosterer/liked_pets_controller.rb]: I think it's nicer for the user to see a pet they have liked has been adopted, rather than just have it disappear from their liked pets, which could be a bit confusing. Therefore, I propose we add the badge if a pet has been adopted. To that end, let's wait until your PR for fosters has been merged, because in order to show the badge we need to determine if the pet has been adopted, and that required the changes to the match, i.e., from singular to plural, on the Pet model.

File: app/controllers/organizations/adopter_fosterer/liked_pets_controller.rb
Comment (kasugaijin) [reply app/controllers/organizations/adopter_fosterer/liked_pets_controller.rb]: https://github.com/rubyforgood/pet-rescue/issues/745

File: app/models/liked_pet.rb
Comment (kasugaijin) [reply app/models/liked_pet.rb]: Done.

**Comment by kasugaijin** [comment]
> I have not looked at the tests yet, but here is my initial review. My main concern is in the comments about the naming of `LikedPet`. Would be curious if others agree.

Yes, I totally agree, and have made the changes to rename as Like instead of LikedPet.

File: app/controllers/organizations/adopter_fosterer/likes_controller.rb
Comment (mononoken) [inline app/controllers/organizations/adopter_fosterer/likes_controller.rb:33]: I have some suggested refactoring here for the policy.

We can remove this `authorize!` line from `#destroy` and move it to `#set_like` but reword it as such:
```ruby
  def set_like
    @like = Like.find(params[:id])
    authorize! @like
  end
```

For actions that affect existing records (such as destroy, update, show), this is the pattern that should usually be used. We fetch the existing record, and then we authorize via that record. Action Policy can infer the correct policy based on the record's class. We don't need the pet context in this case. We should be performing checks directly on the record itself.

We can add this private method to the `LikePolicy`.
```ruby
  def pet
    @pet || record.pet
  end
```
This method will fetch the passed pet context if it exists (so for actions that do not have existing records). However, if the context is not given, it will then check if the passed record responds to `#pet`.

File: test/policies/organizations/like_policy_test.rb
Comment (mononoken) [inline test/policies/organizations/like_policy_test.rb:98]: This is a duplicate name of the above context. Should the tests in this block be moved into the one above?

File: test/policies/organizations/like_policy_test.rb
Comment (mononoken) [inline test/policies/organizations/like_policy_test.rb:179]: This is a duplicate name of the above context. Should the tests in this block be moved into the one above?

**Review by mononoken — APPROVED** [review]
Added a few more comments on the policies and tests. 

Looks good! I think this will be a nice feature for the adopters.

File: app/controllers/organizations/adopter_fosterer/liked_pets_controller.rb
Comment (mononoken) [reply app/controllers/organizations/adopter_fosterer/liked_pets_controller.rb]: I agree that it would probably be confusing for adopters if liked pets just disappeared. Sounds good!

**Pull request merged**
